js/foam/oauth2/GoogleSignIn.js

```diff
@@ -0,0 +1,156 @@
+/**
+ * @license
+ * Copyright 2015 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+CLASS({
+  package: 'foam.oauth2',
+  name: 'GoogleSignIn',
+  requires: [
+    'Binding',
+    'PersistentContext',
+    'foam.dao.IDBDAO',
+    'foam.oauth2.GoogleSignInTokenCarrier',
+    'foam.ui.md.PopupView',
+  ],
+  imports: [
+    'document',
+    'googleClientId',
+    'window',
+  ],
+
+  constants: {
+    AUTH_HEADER: 'X-FOAM-Auth',
+  },
+
+  properties: [
+    {
+      name: 'popupView_',
+    },
+    {
+      name: 'persistentContext_',
+      factory: function() {
+        return this.PersistentContext.create({
+          dao: this.IDBDAO.create({
+            model: this.Binding,
+            name: 'google-auth-' + this.googleClientId
+          }),
+          predicate: NOT_TRANSIENT,
+          context: this
+        });
+      }
+    },
+    {
+      name: 'authToken_',
+    },
+  ],
+
+  methods: [
+    function alogin(ret) {
+      // We might already have a valid Google auth token stored.
+      var future = this.persistentContext_.bindObject('authToken_',
+          this.GoogleSignInTokenCarrier, undefined, 2);
+
+      future(function(token) {
+        if (token && token.token && new Date(token.expiry) > new Date()) {
+          ret();
+          return;
+        }
+
+        // TODO(braden): Investigate the practicality of reimplementing this small
+        // part of the library, to avoid needing to include it.
+
+        // If we're still here, no valid token was found. So first make sure the
+        // Google sign-in library is loaded.
+        // If it's already loading, this.window.__googleSignInListeners will be
+        // defined, and we should add ourselves to it.
+        if (this.window.__googleSignInListeners) {
```

Comment:
Meaning what, for this code? It doesn't care how many windows there are, so long as the callback and `<script>` tag are on the same one. If multiple windows want `GoogleSignIn`, they'll be separate, I suppose.

For the purposes of the WatLobby, it should be fine - only the admin panel window is authenticated if I understand it properly.